Simplify attendee table, add row actions

barryhughes opened the pull request:
Reduce the number of columns in the attendee list table and add row actions.

[C#61992](https://central.tri.be/issues/61992)

tr1b0t commented:
Status|Message|Line
:-----:|-----|-----
:warning: | Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$this' | [Attendees_Table.php:217](https://github.com/moderntribe/event-tickets/blob/8ebee1f3624bb0d43ce68779c07d3252d0314f66/src/Tribe/Attendees_Table.php#L217)

> via jenkins/codesniffer-phpcs

barryhughes commented on src/resources/js/tickets-attendees.js:
```diff
@@ -1,287 +1,290 @@
-jQuery( document ).ready( function( $ ) {
```
Quick note on the changes in this file: it's mostly unchanged, but that may not be immediately apparent from GitHub's diff viewer. What I have done though is switch the structure a little so we can make functions accessible to external code. 

So far, the only function I have actually made public is `count_columns()`, so it can be reused from attendee table code running in Event Tickets Plus.

borkweb replied:
`?w=1` was a life saver

borkweb commented on src/Tribe/Attendees_Table.php:
```diff
@@ -140,30 +109,30 @@ public function column_cb( $item ) {
+		$purchaser_email = empty( $item[ 'purchaser_email' ] ) ? '' : esc_html( $item[ 'purchaser_email' ] );
+
+		return "
+			<div class='purchaser_name'>{$purchaser_name}</div>
```
Maybe it is github being weird, but I see a big gap between `{` and `$` on this line like so:

![screen shot 2016-07-26 at 4 59 50 pm](https://cloud.githubusercontent.com/assets/430385/17155279/6426d10c-5352-11e6-820f-3ab1ae0276d5.png)

barryhughes replied:
Weird ... you mean you see that in GitHub's viewer or if you pull the code locally?

borkweb replied:
GitHub's viewer. Sounds like it is just a rendering issue :D

barryhughes replied:
Yeah that's really weird - I see it now when I go to the _files_ tab - but here in the _conversation_ tab / actually in the source it's fine. Not sure what's going on there.

borkweb replied:
Time to switch to Bazaar.

borkweb commented on src/Tribe/Attendees_Table.php:
```diff
@@ -194,6 +172,25 @@ public function column_order_status( $item ) {
+	public function get_order_id_url( array $item ) {
+		// Backwards compatibility
+		if ( empty( $item['order_id_url'] ) ) {
+			$item['order_id_url'] = esc_url( get_edit_post_link( $item['order_id'], true ) );
```
Perhaps we shouldn't `esc_url()` here and just escape closer to output (which is [already being done](https://github.com/moderntribe/event-tickets/pull/231/files#diff-cf932265bfaec1459c1d705bec1bb62cR159)).

borkweb commented on src/Tribe/Attendees_Table.php:
```diff
@@ -224,6 +230,54 @@ public function column_ticket( $item ) {
+		 * @param string $unique_id A unique string identifier for the ticket.
+		 * @param array  $item      The item entry.
+		 */
+		return apply_filters( 'tribe_events_tickets_attendees_table_attendee_id_column', $unique_id, $item );
```
I almost with we could find a way to make this filter name 45 characters longer.

barryhughes replied:
I toyed with `tribe_events_tickets_attendees_table_attendee_id_column_expected_to_be_int_receives_unique_id_item_array`

borkweb commented on src/resources/js/tickets-attendees.js:
```diff
@@ -1,287 +1,290 @@
-					.find( '.wp-pointer-arrow' ).css({
+var tribe_event_tickets_attendees = tribe_event_tickets_attendees || {};
+
+( function( $, obj ) {
```
`obj` is so much better than the `my` we've used elsewhere.

barryhughes replied:
Yeah I veered away from `my` because it can mean local in some other languages. I also wondered about using `public` but it sounds like it could become an actual keyword in future so might not be the brightest idea.

borkweb replied:
I toyed with `a_local_variable_that_represents_a_global_variable_in_order_to_shorten_the_global_variable_name_but_so_much_for_that_because_this_is_probably_slightly_longer_why_am_i_so_lonely`

barryhughes commented:
Thanks! Updated (also added a set of additional changes, to be a pain and to meet some additional objectives in [#61992](https://central.tri.be/issues/61992)).

borkweb commented on src/resources/js/tickets-attendees.js:
```diff
@@ -1,193 +1,226 @@
+		/**
+		 * Handle "move" requests for individual rows.
+		 */
+		$( '.row-actions' ).find( '.move-ticket' ).click( function( event ) {
```
Rather than bind click events to x elements resulting in x events being bound, I'd suggest this change:

```
$( '.some-parent-maybe-the-table').on( 'click', '.row-actions .move-ticket', function( event ) {
  // do stuff
} );
```

This will result in 1 event being bound on one element but apply to all `.move-ticket` elements.

barryhughes replied:
It look me a moment to get what you mean but yep good call!

borkweb replied:
Event bubbling ftw! There isn't a bubble emoji so you'll have to settle for a water buffalo 🐃

barryhughes replied:
Updated :pray:

borkweb commented:
:yobogoya:
